### ISSUE TITLE
The roles do not appear to filter the hosts to specific tests

### ISSUE THREAD

**Issue by threezerous** [vincentbernat/serverspec-example#5]
Not sure if I followed this correctly.....but my understanding was....if the roles are defined in Rakefile and the hostname meets that role requirement...only tests under that role folder would be executed for that host. Tests under othe role folders would be ignored.
However...in the following case, it appears the tests under conscs are executed for each host in the hosts file.

Rakefile modification for hosts
<code>
# Return all roles of a given host
def roles(host)
  roles = [ "all" ]
  case host
  when /dmcs.+\-con/
    roles << "conscs"
  when /dmas.+\-con/
    roles << "consas"
  when /dmxpl.+\-con/
    roles << "consxpl"
  end
  roles
end
</code>

hosts file
<code>
dmcs12q-con-08
dmcs11q-con-08
dmas11q-con-08
dmas12q-con-08
</code>

Execution output
$ bundle exec rake spec
DL is deprecated, please use Fiddle
user@dmcs12q-con-08's password:
[OK      ] dmcs12q-con-08.
DL is deprecated, please use Fiddle
user@dmcs11q-con-08's password:
[OK      ] dmcs11q-con-08
DL is deprecated, please use Fiddle
user@dmas11q-con-08's password:
[33 /37  ] dmas11q-con-08
DL is deprecated, please use Fiddle
user@dmas12q-con-08's password:
[33 /37  ] dmas12q-con-08

I tried running bundle exec rake check:role:conscs, but it fails after the second host as

$ bundle exec rake check:role:conscs
DL is deprecated, please use Fiddle
user@dmcs12q-con-08's password:
[OK      ] dmcs12q-con-08
DL is deprecated, please use Fiddle
user@dmcs11q-con-08's password:
[OK      ] dmcs11q-con-08
rake aborted!
Errno::EINVAL: Invalid argument @ rb_sysopen - ./reports/check:role:conscs--2015
-07-10T14-06-28.json
c:/Users/user/cookbooks/documentum-serverspec-scripts/Rakefile:167:in `initia
lize'
c:/Users/user/cookbooks/documentum-serverspec-scripts/Rakefile:167:in `open'
c:/Users/user/cookbooks/documentum-serverspec-scripts/Rakefile:167:in `block
(2 levels) in <top (required)>'
c:/Users/user/cookbooks/documentum-serverspec-scripts/Rakefile:210:in `block
in <top (required)>'
Tasks: TOP => reports:build
(See full trace by running task with --trace)

Is it possible that my regex is not evaluating all hosts in the rakefile in roles section?

**Comment by vincentbernat** [comment]
Your `roles` function seem fine (but you don't need to escape `-`, it hasn't any special meaning when not inside brackets).

Could you add a `puts t.target, t.pattern` after line 104 to check if the pattern is correct. If it's correct, maybe that's the antislash used by Windows for paths. Could you try to replace the `t.pattern = ...` line with:

```ruby
t.pattern = './spec/{' + dirs.join(",") + '}/*_spec.rb')
```

And check if it changes anything. Maybe the antislash makes the pattern invalid.

For the other error, it's again because of the `:` in path, I'll push a commit for that.

**Comment by vincentbernat** [comment]
Line 104 is the line after `t.pattern = ...`. It is unlikely to be 104 for you.

**Comment by vincentbernat** [comment]
See 3ae163802e62 for the last error.

**Comment by threezerous** [comment]
Thank you for quick attention on this
Adding ```puts t.target, t.pattern\\\ after original t.pattern line gave the following output, which appears correct
```
./spec/{all,conscs,dmcs11q-con-08}/*_spec.rb
./spec/{all,consas,dmas11q-con-08}/*_spec.rb
```

Changing the t.pattern line  to
```
t.pattern = File.join('./spec/{' + dirs.join(",") + '}/*_spec.rb')
```
didn't help since the ```bundle exec rake spec``` still navigated through all hosts, while
```bundle exec rake check:role:conscs``` gave same errors.

**Comment by vincentbernat** [comment]
For `bundle exec rake check:role:conscs`, you need the commit that I mentioned in the previous comment.

You also say that `bundle exec rake spec` navigates through all hosts, but this is expected. You mean all tests I assume.

What does `bundle list` says?

**Comment by vincentbernat** [comment]
Also, try `env TARGET_HOST=dmcs12q-con-08 bundle exec rspec --pattern "./spec/{all,conscs,dmcs11q-con-08}/*_spec.rb"`. It should only run tests in the specified directories.

**Comment by threezerous** [comment]
Awesome !! The latest commit fixed the ```bundle exec rake check:role:conscs``` errors.
Regarding ```bundle exec rake spec``` I have only one test to meet one role conscs. But I have two hosts that do not meet that role condition (the hostnames start with dmas)......so I would have expected ```bundle exec rake spec``` not to run the tests under conscs to run for the dmas hosts. But if you think, it will navigate through all hosts for each test for ```bundle exec rake spec```, I am fine with it. The command ```bundle exec rake check:role:conscs``` does address my requirement

**Comment by vincentbernat** [comment]
Yes, the tests under `conscs` should not be run for the `dmas` hosts.

Try:

```
env TARGET_HOST=dmas12q-con-08 bundle exec rspec --pattern "./spec/{all,consas,dmas12q-con-08}/*_spec.rb"
```

The tests in `conscs` should not be executed. If they execute, this is a bug in rspec. If they don't, check they don't run either with `bundle exec rake check:server:dmas12q-con-08`.

**Comment by threezerous** [comment]
The tests do not get executed with 
```env TARGET_HOST=dmas12q-con-08 bundle exec rspec --pattern "./spec/{all,consas,dmas12q-con-08}/*_spec.rb"```

but do run with
```bundle exec rake check:server:dmas12q-con-08```

**Comment by vincentbernat** [comment]
Well, that's odd. I need to look a bit at at Rake to understand why it would work with `spec`, but not when using the lib.

**Comment by threezerous** [comment]
Another windows related question, if I may :(
Looks like the ASK_SUDO_PASSWORD does not seem to be recognized on windows git bash...though I confirmed the variable
![image](https://cloud.githubusercontent.com/assets/4130145/8784549/a907c7cc-2ef0-11e5-97f5-14ad8e5be712.png)
 is set

**Comment by vincentbernat** [comment]
The message is a bit misleading. If you want to use a non-password-less sudo, you need to provide it as `c.sudo_password`. The original snippet used as `spec_helper.rb` was:

```ruby
RSpec.configure do |c|
  if ENV['ASK_SUDO_PASSWORD']
    require 'highline/import'
    c.sudo_password = ask("Enter sudo password: ") { |q| q.echo = false }
  else
    c.sudo_password = ENV['SUDO_PASSWORD']
  end
# ...
end
```

**Comment by threezerous** [comment]
That does not seem to make a difference. I tried both prompting for a password (ASK_SUDO_PASSWORD=1) or hard-codong the password in the spec_helper.rb. See screen-capture
![image](https://cloud.githubusercontent.com/assets/4130145/8786347/904601da-2efb-11e5-83d1-f69f5541eed2.png)

**Comment by vincentbernat** [comment]
You seem to hit a case where sudo complains there is no TTY available: https://github.com/serverspec/specinfra/blob/v1/lib/specinfra/backend/ssh.rb#L88. Try setting `request_pty` as well to see if it helps.

**Comment by threezerous** [comment]
That was it !...Thanks
